src/wallet/db.cpp

```diff
@@ -402,11 +403,41 @@ CDB::CDB(CWalletDBWrapper& dbw, const char* pszMode, bool fFlushOnCloseIn) : pdb
                             0);
 
             if (ret != 0) {
+                error = strprintf("CDB: Error %d, can't open database %s", ret, strFilename);
+            } else {
+                // Make sure database has a unique fileid. If it doesn't, throw
+                // an error. BDB caches do not work properly when more than one
+                // open database has the same fileid (values written to one
+                // database may show up in reads to other databases).
+                //
+                // Bitcoin will never create different databases with the same
+                // fileid, but this error can be triggered if users manually
+                // copy database files.
+                u_int8_t fileid[DB_FILE_ID_LEN];
+                ret = pdb->get_mpf()->get_fileid(fileid);
+                if (ret != 0) {
+                    error = strprintf("CDB: Can't open database %s (get_fileid error %d)\n", strFilename, ret);
+                }
+
+                for (const auto& item : env->mapDb) {
+                    u_int8_t item_fileid[DB_FILE_ID_LEN];
+                    if (item.second && item.second->get_mpf()->get_fileid(item_fileid) == 0 &&
+                        memcmp(fileid, item_fileid, sizeof(fileid)) == 0) {
+                        const char* item_filename = nullptr;
+                        item.second->get_dbname(&item_filename, nullptr);
+                        error = strprintf("CDB: Can't open database %s (duplicates fileid %s from %s)", strFilename,
+                            HexStr(std::begin(item_fileid), std::end(item_fileid)),
+                            item_filename ? item_filename : "");
+                    }
+                }
+            }
+
+            if (!error.empty()) {
                 delete pdb;
                 pdb = nullptr;
                 --env->mapFileUseCount[strFile];
                 strFile = "";
-                throw std::runtime_error(strprintf("CDB: Error %d, can't open database %s", ret, strFilename));
+                throw std::runtime_error(error);
```

Comment:
I guess the reason for storing the error in a string, and not just raising the exceptions at the place the error occurs, above, is to avoid duplicating this cleanup code?
It seems a bit circuitous, but I understand. An alternative would be to wrap these operations in a RAII operation.

Reply:
This was one reason. The other reason was wanting this to be self-contained and not change existing code.